### ISSUE TITLE
Why don't the protocol codes simply start at 0 and increment?

### ISSUE THREAD

**Issue by aunyks** [multiformats/multiaddr#58]
In `protocols.csv`, I see that the `code` column contains the integer representations of each protocol under multiaddress. However, I do not notice a pattern between each protocol and its respective code. Why don't the codes just start at 0 and increment for each new protocol added? Is there a pattern or set of rules that constitutes the assignment of a code number to a protocol?

**Comment by Stebalien** [comment]
There are some patterns but many of the choices exist for legacy reasons. We used to have multiple tables (one for multiaddrs, one for IPLD formats, one for multihashes, one for multibases, etc.) but have now unified them in the https://github.com/multiformats/multicodec repo (we actually need to remove the protocols.csv table from this repo). However, we decided to just unify them (and luckily didn't have any collisions).

Other patterns:

* Keep some single byte (<127) multicodecs available.
* (try to) Keep related multicodecs next to each other.
* Many multiaddrs were chosen based on the associated port number. However, some of these port numbers conflicted with multicodecs we'd already chosen so we had to choose alternates.

TL;DR: It's kind of a mess.
